biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/Activator.java

```diff
@@ -0,0 +1,377 @@
+			}
+		});
+		summary = new SummaryGeneratingListener();
+		listenerList.add(basicReport);
+		listenerList.add(summary);
+		listeners = listenerList.stream()
+			.toArray(TestExecutionListener[]::new);
+
+		if (testcases == null) {
+			trace("automatic testing of all bundles with " + aQute.bnd.osgi.Constants.TESTCASES + " header");
+			try {
+				automatic();
+			} catch (IOException e) {
+				// ignore
+			}
+		} else {
+			trace("receivednames of classes to test %s", testcases);
+			try {
+				baseSelectors = Stream.of(testcases.split("\\s*,\\s*"))
+					.map(x -> {
+						if (x.contains(":")) {
+							return selectMethod(x.replace(':', '#'));
+						} else {
+							return selectClass(x);
+						}
+					})
+					.collect(Collectors.toList());
+				automatic();
+			} catch (Exception e) {
+				e.printStackTrace();
+				System.exit(254);
+			}
+		}
+	}
+
+	void automatic() throws IOException {
+		final BlockingDeque<LauncherDiscoveryRequest> queue = new LinkedBlockingDeque<>();
+
+		trace("adding Bundle Listener for getting test bundle events");
+		context.addBundleListener((SynchronousBundleListener) event -> {
+			switch (event.getType()) {
+				case BundleEvent.STARTED :
+				case BundleEvent.RESOLVED :
+					checkBundle(queue, event.getBundle());
+					break;
+			}
+		});
+
+		long result = 0;
+		LauncherDiscoveryRequest request = buildRequest();
+		TestPlan plan = launcher.discover(request);
+		if (plan.containsTests()) {
+			result = test(request);
+			if (!continuous) {
+				System.exit((int) result);
+			}
+		}
+
+		trace("starting queue");
+		while (active()) {
+			try {
+				LauncherDiscoveryRequest testRequest = queue.takeFirst();
+				trace("test will run");
+				result += test(testRequest);
+				trace("test ran");
+				if (queue.isEmpty() && !continuous) {
+					trace("queue " + queue);
+					System.exit((int) result);
+				}
+			} catch (InterruptedException e) {
+				trace("tests bundle queue interrupted");
+				thread.interrupt();
+			} catch (Exception e) {
+				error("Not sure what happened anymore %s", e);
+				System.exit(254);
+			}
+		}
+	}
+
+	List<DiscoverySelector> baseSelectors = Collections.emptyList();
+
+	LauncherDiscoveryRequest buildRequest() {
+		return LauncherDiscoveryRequestBuilder.request()
+			.configurationParameter(BundleEngine.CHECK_UNRESOLVED, unresolved)
+			.selectors(baseSelectors)
+			.build();
+	}
+
+	LauncherDiscoveryRequest buildRequest(Bundle bundle) {
+		List<DiscoverySelector> selectors = new ArrayList<>(baseSelectors.size() + 1);
+		selectors.add(BundleSelector.selectBundle(bundle));
+		return LauncherDiscoveryRequestBuilder.request()
+			.configurationParameter(BundleEngine.CHECK_UNRESOLVED, unresolved)
+			.selectors(selectors)
+			.build();
+	}
+
+	void checkBundle(BlockingDeque<LauncherDiscoveryRequest> queue, Bundle bundle) {
+		Bundle host = BundleUtils.getHost(bundle)
+			.orElse(bundle);
+		if (host.getState() == Bundle.ACTIVE || host.getState() == Bundle.STARTING) {
```

Comment:
Bundle state checks are best written as bit mask tests.

```java
if ((host.getState() & (Bundle.ACTIVE | Bundle.STARTING)) != 0) {
```

Reply:
Done. That's clearly more efficient if checking multiple states and possibly more readable too - thanks for the tip!